core/stat/base/slidingwindow_metric.go

```diff
@@ -0,0 +1,150 @@
+package base
+
+import (
+	"fmt"
+	"github.com/sentinel-group/sentinel-golang/core/base"
+	"github.com/sentinel-group/sentinel-golang/util"
+	"math"
+)
+
+// The basic metric struct,
+// SlidingWindowMetric doesn't store any data and is the wrapper of BucketLeapArray to adapt to different internal window
+// SlidingWindowMetric is used for SentinelRules and BucketLeapArray is used for monitor
+// BucketLeapArray is per resource, and SlidingWindowMetric support only read operation.
+type SlidingWindowMetric struct {
+	windowLengthInMs uint32
+	sampleCount      uint32
+	intervalInMs     uint32
+	real             *BucketLeapArray
+}
+
+// It must pass the parameter point to the real storage entity
+func NewSlidingWindowMetric(intervalInMs, sampleCount uint32, real *BucketLeapArray) *SlidingWindowMetric {
+	if real == nil || intervalInMs <= 0 || sampleCount <= 0 {
+		panic(fmt.Sprintf("Illegal parameters,intervalInMs=%d,sampleCount=%d,real=%+v.\n", intervalInMs, sampleCount, real))
+	}
+
+	if intervalInMs%sampleCount != 0 {
+		panic(fmt.Sprintf("Invalid parameters, intervalInMs is %d, sampleCount is %d.\n", intervalInMs, sampleCount))
+	}
+	winLengthInMs := intervalInMs / sampleCount
+
+	parentIntervalInMs := real.IntervalInMs()
+	parentWindowLengthInMs := real.WindowLengthInMs()
+
+	if winLengthInMs != parentWindowLengthInMs {
+		panic(fmt.Sprintf("SlidingWindowMetric's WindowLengthInMs(%d) is not equal to real BucketLeapArray's WindowLengthInMs(%d).\n", winLengthInMs, parentWindowLengthInMs))
+	}
+
+	if intervalInMs > parentIntervalInMs {
+		// todo if SlidingWindowMetric's intervalInMs is greater than BucketLeapArray.
+		panic(fmt.Sprintf("The length(%d) of SlidingWindowMetric is greater than parent BucketLeapArray(%d).\n", intervalInMs, parentIntervalInMs))
+	}
+
+	// 10 * 1000 ms == parent
+	if parentIntervalInMs%intervalInMs != 0 {
+		panic(fmt.Sprintf("SlidingWindowMetric's intervalInMs(%d) is not divisible by real BucketLeapArray's intervalInMs(%d).\n", intervalInMs, parentIntervalInMs))
+	}
+
+	return &SlidingWindowMetric{
+		windowLengthInMs: winLengthInMs,
+		sampleCount:      sampleCount,
+		intervalInMs:     intervalInMs,
+		real:             real,
+	}
+}
+
+// Get the [start index, end index) in BucketLeapArray for SlidingWindowMetric in current time.
+func (m *SlidingWindowMetric) getIndexInParent(now uint64) (startIdx, endIdx int) {
```

Comment:
Any test cases for this method?
Assuming the leap array `a(interval=ia, sampleCount=sa)` and the wrapper SlidingWindowMetric `m(interval=im, sampleCount=sm)` which share the same `bucketLength` value (in fact the bucketLength can be different). The actual end idx should be `a.calculateTimeIdx(currentTime)` and the start idx should be `endIdx - sm`). For example, if sa=20, sm=10, idx in LeapArray for current time is 4, then the actual range of the values should be (4-10, 4], i.e. `[0,4]+[15,19]`.

Reply:
Had add unit test.